community/cypher/cypher-planner-3.4/src/main/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/ReportCostComparisonsAsRows.scala

```diff
@@ -0,0 +1,194 @@
+  * that looks something like this:
+  *
+  * UNWIND [{comparison: 0, planId: 1, planText: "plan", planCost: "costs", cost: 12, ...}] as col
+  * RETURN col["comparison] as comparison
+  *
+  * This is the plan that actually gets run, and not the one that the user provided.
+  *
+  * WARNING: This is a debug feature, and as such, not tested as rigorously as other features are.
+  *          Only run this on a system that is not live after taking proper backups!
+  *
+  */
+class ReportCostComparisonsAsRows extends CostComparisonListener {
+
+  private val rows = new mutable.ListBuffer[Row]()
+  private var comparisonCount = 0
+  private val pos = InputPosition(0, 0, 0)
+
+  override def report[X](projector: X => LogicalPlan,
+                         input: Iterable[X],
+                         inputOrdering: Ordering[X],
+                         context: LogicalPlanningContext,
+                         solveds: Solveds,
+                         cardinalities: Cardinalities): Unit = if (input.size > 1) {
+
+    def stringTo(level: Int, plan: LogicalPlan): String = {
+      def indent(level: Int, in: String): String = level match {
+        case 0 => in
+        case _ => "\n" + "  " * level + in
+      }
+
+      val cost = context.cost(plan, context.input, cardinalities)
+      val thisPlan = indent(level, s"${plan.getClass.getSimpleName} costs $cost cardinality ${cardinalities.get(plan.id)}")
+      val l = plan.lhs.map(p => stringTo(level + 1, p)).getOrElse("")
+      val r = plan.rhs.map(p => stringTo(level + 1, p)).getOrElse("")
+      thisPlan + l + r
+    }
+
+    val sortedPlans = input.toIndexedSeq.sorted(inputOrdering).map(projector).reverse
+    val winner = sortedPlans.last
+
+    val theseRows: immutable.Seq[Row] = sortedPlans.map { plan =>
+      val planText = plan.toString.replaceAll(System.lineSeparator(), System.lineSeparator())
+      val planTextWithCosts = stringTo(0, plan).replaceAll(System.lineSeparator(), System.lineSeparator())
+      val cost = context.cost(plan, context.input, cardinalities)
+      val cardinality = cardinalities.get(plan.id)
+
+      Row(comparisonCount, plan.id, planText, planTextWithCosts, cost, cardinality, winner.id == plan.id)
+    }
+
+    rows ++= theseRows
+    comparisonCount += 1
+  }
+
+  def addPlan(in: LogicalPlanState): LogicalPlanState = {
+    val plan = asPlan()
+    val newStatement = asStatement()
+    val solveds = new Solveds
+    val cardinalities = new Cardinalities
+
+    var current: Option[LogicalPlan] = Some(plan)
+    do {
+      val thisPlan = current.get
+      solveds.set(thisPlan.id, PlannerQuery.empty)
+      cardinalities.set(thisPlan.id, Cardinality.SINGLE)
+      current = current.get.lhs
+    } while (current.nonEmpty)
+
+    in.copy(maybePeriodicCommit = Some(None), maybeLogicalPlan = Some(plan), maybeStatement = Some(newStatement), solveds = solveds, cardinalities = cardinalities)
```

Comment:
Maybe not related but `Some(None)` just looks wrong to me

Reply:
I know it's weird... The outer `Option` signifies whether or not the field is set. The inner `Option` signifies whether we have a periodic commit or not. It sure looks wrong, but it's right in this particular instance.